controllers/managedfusiondeployment_controller.go

```diff
@@ -0,0 +1,446 @@
+	alertmanagerConfigSecretName         = "managed-fusion-alertmanager-config"
+	agentCSVPrefix                       = "managed-fusion-agent"
+	monLabelKey                          = "app"
+	monLabelValue                        = "managed-fusion-agent"
+	pagerDutyKey                         = "pagerDutyServiceKey"
+	smtpPasswordKey                      = "smtpAuthPassword"
+)
+
+type ManagedFusionDeploymentReconciler struct {
+	Client             client.Client
+	UnrestrictedClient client.Client
+	Log                logr.Logger
+	Scheme             *runtime.Scheme
+	Namespace          string
+
+	ctx                           context.Context
+	prometheus                    *promv1.Prometheus
+	alertmanager                  *promv1.Alertmanager
+	alertmanagerConfig            *promv1a1.AlertmanagerConfig
+	alertmanagerConfigSecret      *corev1.Secret
+	managedFusionDeploymentSecret *corev1.Secret
+	smtpConfigData                *smtpConfig
+	pagerDutyConfigData           *pagerDutyConfig
+	CustomerNotificationHTMLPath  string
+}
+
+// Add necessary rbac permissions for managedfusiondeployment finalizer in order to set blockOwnerDeletion.
+// +kubebuilder:rbac:groups="monitoring.coreos.com",resources={alertmanagers,prometheuses,alertmanagerconfigs},verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="monitoring.coreos.com",resources=prometheusrules,verbs=get;list;watch;create;update;patch
+// +kubebuilder:rbac:groups="monitoring.coreos.com",resources=podmonitors,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups="monitoring.coreos.com",resources=servicemonitors,verbs=get;list;watch;update;patch;create;delete
+// +kubebuilder:rbac:groups="",resources={secrets,secrets/finalizers},verbs=create;get;list;watch;update
+// +kubebuilder:rbac:groups=operators.coreos.com,resources=clusterserviceversions,verbs=get;list;watch;delete;update;patch
+// +kubebuilder:rbac:groups="apps",resources=statefulsets,verbs=get;list;watch
+// +kubebuilder:rbac:groups="coordination.k8s.io",resources=leases,verbs=create;get;list;watch;update
+
+// SetupWithManager creates an setup a ManagedFusionDeploymentReconciler to work with the provided manager
+func (r *ManagedFusionDeploymentReconciler) SetupWithManager(mgr ctrl.Manager, ctrlOptions *controller.Options) error {
+	if ctrlOptions == nil {
+		ctrlOptions = &controller.Options{
+			MaxConcurrentReconciles: 1,
+		}
+	}
+	// Filter the events for objects that are not owned by managed-fusion-agent-config secret
+	filterNonOwnedEvents := predicate.NewPredicateFuncs(
+		func(object client.Object) bool {
+			if object.GetName() == managedFusionAgentSecretName &&
+				object.GetNamespace() == r.Namespace {
+				return true
+			} else {
+				for _, owner := range object.GetOwnerReferences() {
+					if owner.Kind == "Secret" &&
+						owner.Name == managedFusionAgentSecretName {
+						return true
+					}
+				}
+			}
+			return false
```

Comment:
For us that's not the case, we have secrets object in FOR watch and we have secret object in OWN watch so I thought it is better to have a event filter that will enqueue requests for managed fusion secret and for the objects that are owned by managed fusion secret. This will be done using filterNonOwnedEvents predicate.

Reply:
But these are different use cases and one will be covered in the event redirect.

For the input secret, we will have a `managedFusionSecretPredicates` that filter for our `managedFusionSecretName` in our namespace 

For other secrets we will not have a predicate at all and will use an Owns watch, that will filter for only secrets that are owned by us. 
 
And please remove the global `WithEventFilter` as it have unwanted side effects

Reply:
Updated in latest commit